packages/admin/cms-admin/src/form/file/FinalFormFileUpload.tsx

```diff
@@ -0,0 +1,175 @@
+import { gql } from "@apollo/client";
+import { ErrorFileSelectItem, FileSelect, FileSelectProps, LoadingFileSelectItem } from "@comet/admin";
+import { saveAs } from "file-saver";
+import React from "react";
+import { FieldRenderProps } from "react-final-form";
+import { FormattedMessage } from "react-intl";
+
+import { GQLFinalFormFileUploadFragment } from "./FinalFormFileUpload.generated";
+
+export const finalFormFileUploadFragment = gql`
+    fragment FinalFormFileUpload on PublicUpload {
+        id
+        name
+        size
+    }
+`;
+
+type SuccessfulApiResponse = {
+    id: string;
+    createdAt: string;
+    updatedAt: string;
+    name: string;
+    size: number;
+    mimetype: string;
+    contentHash: string;
+};
+
+type FailedApiResponse = {
+    statusCode: number;
+    message: string;
+};
+
+type ApiResponse = SuccessfulApiResponse | FailedApiResponse;
+
+export type FinalFormFileUploadProps<MaxFiles> = (MaxFiles extends 1
+    ? { maxFiles?: MaxFiles } & FieldRenderProps<GQLFinalFormFileUploadFragment, HTMLInputElement>
+    : { maxFiles?: MaxFiles } & FieldRenderProps<GQLFinalFormFileUploadFragment[], HTMLInputElement>) &
+    Partial<FileSelectProps<GQLFinalFormFileUploadFragment>>;
+
+export const FinalFormFileUpload = <MaxFiles extends number | undefined>({ input, maxFiles, ...restProps }: FinalFormFileUploadProps<MaxFiles>) => {
+    const [tooManyFilesSelected, setTooManyFilesSelected] = React.useState(false);
+    const [uploadingFiles, setUploadingFiles] = React.useState<LoadingFileSelectItem[]>([]);
+    const [failedUploads, setFailedUploads] = React.useState<ErrorFileSelectItem[]>([]);
+    const [downloadingFileIds, setDownloadingFileIds] = React.useState<string[]>([]);
+    const [failedToDownloadFileIds, setFailedToDownloadFileIds] = React.useState<string[]>([]);
+
+    const apiUrl = "http://localhost:4000"; // TODO: Where do we get the url from? Env? Hook?
```

Comment:
Good question. Maybe we need a PublicUploads Context for this. @nsams what do you think?

Reply:
- for dam we have a damConfig (part of CmsBlockContext) that has apiUrl and apiClient (=a axios instance configured with apiUrl as baseURL)
    - that is too dam specific to be used here
- we also have ApolloProvider containing the apolloClient which has the httpLink with apiUrl set
    - we need the apiUrl without `/graphql`
    - you probably can not easily access that url
- we have the config.apiUrl, but that is not injected into React
- we have the API_URL env var, but that should not be used (use properly typed and validated config instead)

Ideas:
- create a PublicUplodas Context
- create a more generic ApiUrl context (that can potentially be used for dam as well)
- set the apiUrl as prop to the FileUpload field (in the application by importing config)

---

BTW: This is a FinalFormFileUpload component, should that always use public uplodas as upload target?

Reply:
> create a more generic ApiUrl context (that can potentially be used for dam as well)

That sounds like the best option to me. 

Or would it make sense to just move `apiUrl` outside of `damConfig` in `CmsBlockContext`?
Then we don't need two contexts but the name `CmsBlock` might not be fitting anymore, as it's used in a non-Block component 🤷🏼‍♂️ 

```diff
 export interface CmsBlockContext {
+    apiUrl: string;
     damConfig: {
-        apiUrl: string;
         // ...
     };
     // ...
 }
```

> BTW: This is a FinalFormFileUpload component, should that always use public uplodas as upload target?

I think public uploads should at least be the default. 
As soon as we have a use-case where we need another target, we could add support for this. 

> set the apiUrl as prop to the FileUpload field (in the application by importing config)

I think this would make sense, if we commonly need a custom upload target but I don't think that will be the case. 

---

@johnnyomair what do you think?

Reply:
> we have the config.apiUrl, but that is not injected into React

We have a ConfigContext in the Starter, maybe we could convert that into a CometConfig context in the library (instead of the ApiUrl context)?

> Or would it make sense to just move apiUrl outside of damConfig in CmsBlockContext?

CmsBlockContext is very specific to the block methods, I wouldn't change that (we may accidentally have done that already)

Reply:
> We have a ConfigContext in the Starter, maybe we could convert that into a CometConfig context in the library (instead of the ApiUrl context)?

That sounds like a good idea. Should we do this?

Should we maybe use `damConfig.apiUrl` as a temporary solution and create/discuss the permanent solution in a separate PR?

Reply:
Temporary solution done in 6d44a6ae50e5c943a45532b444265daccbbc186c.

Also created a task to find a better/permanent solution: COM-953.
_(FYI @thomasdax98)_